common/core/desktop/tests/unit/kmx/kmx.cpp

```diff
@@ -337,12 +353,37 @@ int run_test(const km::kbp::path &source, const km::kbp::path &compiled) {
     if (p.vk == KM_KBP_VKEY_CAPS) {
       toggle_caps_lock_state();
     }
+
     for (auto key_down = 1; key_down >= 0; key_down--) {
       try_status(km_kbp_process_event(test_state, p.vk, p.modifier_state | caps_lock_state(), key_down));
+
       for (auto act = km_kbp_state_action_items(test_state, nullptr); act->type != KM_KBP_IT_END; act++) {
-        apply_action(test_state, *act, text_store, options);
+        apply_action(test_state, *act, text_store, test_context, options);
       }
     }
+
+    // Compare context and text store at each step - should be identical
+    size_t n = 0;
+    try_status(km_kbp_context_get(km_kbp_state_context(test_state), &citems));
+    try_status(km_kbp_context_items_to_utf16(citems, nullptr, &n));
+    km_kbp_cp *buf = new km_kbp_cp[n];
+    try_status(km_kbp_context_items_to_utf16(citems, buf, &n));
+
+    // Verify that both our local test_context and the core's test_state.context have
+    // not diverged
+    auto ci = citems;
+    for(auto test_ci = test_context.begin(); ci->type != KM_KBP_CT_END || test_ci != test_context.end(); ci++, test_ci++) {
```

Comment:
Shouldn't this be `&&`?  Same below in line 402.

```
for(auto test_ci = test_context.begin(); ci->type != KM_KBP_CT_END && test_ci != test_context.end(); ci++, test_ci++) {
```

Reply:
No, because otherwise if one list is longer than the other, we don't assert that. The following line asserts that the lists are the same length effectively (harder to do outside the loop because `test_ci` is loop-local).